[Writing Tools] Outlook.app quits unexpectedly with Writing Tools

rr-codes opened the pull request:
#### 48d994887c9456fe1f7ea44a6d0859fa07eb266b
<pre>
[Writing Tools] Outlook.app quits unexpectedly with Writing Tools
<a href="https://bugs.webkit.org/show_bug.cgi?id=283098">https://bugs.webkit.org/show_bug.cgi?id=283098</a>
<a href="https://rdar.apple.com/139513909">rdar://139513909</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

The `SystemFallbackFontCache` system depends on using a `WebCore::Timer`. In the destructor of `Font`,
a font cache is accessed and the font is removed from the cache. Notably, even if there is no cache,
one is erroneously created anyways.

The `AttributedString` type uses the `Font` type and therefore indirectly uses a `WebCore::Timer`.
Therefore, any use of `AttributedString` in the UI process is unsafe, as it is not allowed to use a
`WebCore::Timer` in the UI process.

This issue manifests in the implementation of Writing Tools, which uses AttributedStrings in the UI
process. This issue was always present, however it turned into a crash once a release assert was a
dded in `WebCore::Timer`.

To fix this, ensure that a new font cache is not created in the destructor of Font.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::~Font):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::forCurrentThreadIfExists):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, AttributedStringWithWebKitLegacy)):

Canonical link: <a href="https://commits.webkit.org/286579@main">https://commits.webkit.org/286579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f4e2ace87cc41e713b8948b8502b237e973929f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76404 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29310 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80925 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27680 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78521 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64581 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3732 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59901 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18023 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79471 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49817 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65612 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40233 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47212 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23101 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26002 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23433 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82377 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3780 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/2473 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68118 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3933 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65581 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67432 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16816 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11398 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9498 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3727 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6536 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3750 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7180 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5508 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -3889,4 +3889,26 @@ InstanceMethodSwizzler swizzler(PAL::getWTWritingToolsClass(), @selector(showToo
 #endif
 }
 
+#if PLATFORM(IOS_FAMILY) && !ASSERT_ENABLED
```
Why only for release builds?

rr-codes committed:
[Writing Tools] Outlook.app quits unexpectedly with Writing Tools
https://bugs.webkit.org/show_bug.cgi?id=283098
rdar://139513909

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

The `SystemFallbackFontCache` system depends on using a `WebCore::Timer`. In the destructor of `Font`,
a font cache is accessed and the font is removed from the cache. Notably, even if there is no cache,
one is erroneously created anyways.

The `AttributedString` type uses the `Font` type and therefore indirectly uses a `WebCore::Timer`.
Therefore, any use of `AttributedString` in the UI process is unsafe, as it is not allowed to use a
`WebCore::Timer` in the UI process.

This issue manifests in the implementation of Writing Tools, which uses AttributedStrings in the UI
process. This issue was always present, however it turned into a crash once a release assert was a
dded in `WebCore::Timer`.

To fix this, ensure that a new font cache is not created in the destructor of Font.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::~Font):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.cpp:
(WebCore::SystemFallbackFontCache::forCurrentThreadIfExists):
* Source/WebCore/platform/graphics/SystemFallbackFontCache.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, AttributedStringWithWebKitLegacy)):

Canonical link: https://commits.webkit.org/286579@main
